File: 872_leaf_similar_nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def leafSimilar(self, root1: TreeNode | None, root2: TreeNode | None) -> bool:
        values1 = self._get_leaf_values(root1)
        values2 = self._get_leaf_values(root2)
        return values1 == values2
    
    def _get_leaf_values(self, root: TreeNode | None) -> list[int]:
        values = []
        stack: list[TreeNode | None] = []
        cur = root
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            
            cur = stack.pop()
            is_leaf = cur.left is None and cur.right is None
            if is_leaf:
                values.append(cur.val)

            cur = cur.right
        return values

File: problems/872-leaf-similar-nodes.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function leafSimilar(root1: TreeNode | null, root2: TreeNode | null): boolean {
  const leafNums1 = getLeafNums(root1)
  const leafNums2 = getLeafNums(root2)
  if (leafNums1.length !== leafNums2.length) {
    return false
  }
  for (let i = 0; i < leafNums1.length; i++) {
    if (leafNums1[i] !== leafNums2[i]) {
      return false
    }
  }
  return true
}

function getLeafNums(root: TreeNode | null): number[] {
  const nums: number[] = []
  getLeafNumsHelper(root, nums)
  return nums
}

function getLeafNumsHelper(node: TreeNode | null, nums: number[]): void {
  if (!node) {
    return
  }
  const isLeaf = node.left === null && node.right === null
  getLeafNumsHelper(node.left, nums)
  if (isLeaf) {
    nums.push(node.val)
  }
  getLeafNumsHelper(node.right, nums)
}
